File: WEEK7/Js/bj12_2162.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const lines = [];
for (let i = 1; i <= n; i++) {
    lines.push(input[i].split(' ').map(BigInt));
}

const parent = Array.from({ length: n }, (_, i) => i);
const size = new Int32Array(n).fill(1);

function find(x) {
    if (parent[x] === x) return x;
    return parent[x] = find(parent[x]);
}

function union(a, b) {
    const rootA = find(a);
    const rootB = find(b);
    if (rootA !== rootB) {
        parent[rootB] = rootA;
        size[rootA] += size[rootB];
    }
}

function ccw(ax, ay, bx, by, cx, cy) {
    const res = (bx - ax) * (cy - ay) - (cx - ax) * (by - ay);
    if (res > 0n) return 1;
    if (res < 0n) return -1;
    return 0;
}

function check(l1, l2) {
    const [x1, y1, x2, y2] = l1;
    const [x3, y3, x4, y4] = l2;

    const ans1 = ccw(x1, y1, x2, y2, x3, y3);
    const ans2 = ccw(x1, y1, x2, y2, x4, y4);
    const ans3 = ccw(x3, y3, x4, y4, x1, y1);
    const ans4 = ccw(x3, y3, x4, y4, x2, y2);

    if (ans1 * ans2 === 0 && ans3 * ans4 === 0) {
        const [minX1, maxX1] = x1 < x2 ? [x1, x2] : [x2, x1];
        const [minX3, maxX3] = x3 < x4 ? [x3, x4] : [x4, x3];
        const [minY1, maxY1] = y1 < y2 ? [y1, y2] : [y2, y1];
        const [minY3, maxY3] = y3 < y4 ? [y3, y4] : [y4, y3];
        return minX1 <= maxX3 && minX3 <= maxX1 && minY1 <= maxY3 && minY3 <= maxY1;
    }
    return ans1 * ans2 <= 0 && ans3 * ans4 <= 0;
}

for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
        if (check(lines[i], lines[j])) {
            union(i, j);
        }
    }
}

let groupCount = 0;
let maxSize = 0;

for (let i = 0; i < n; i++) {
    if (parent[i] === i) {
        groupCount++;
        if (size[i] > maxSize) maxSize = size[i];
    }
}

console.log(groupCount);
console.log(maxSize);

File: WEEK7/Python/bj12_2162.py
import sys


def find(x):
    if parent[x] == x:
        return x
    parent[x] = find(parent[x])
    return parent[x]


def union(a, b):
    rootA = find(a)
    rootB = find(b)
    if rootA != rootB:
        parent[rootB] = rootA
        size[rootA] += size[rootB]


def ccw(x1, y1, x2, y2, x3, y3):
    temp = (x2 - x1) * (y3 - y1) - (x3 - x1) * (y2 - y1)
    if temp > 0:
        return 1
    elif temp < 0:
        return -1
    return 0


def check(l1, l2):
    x1, y1, x2, y2 = l1
    x3, y3, x4, y4 = l2

    ans1 = ccw(x1, y1, x2, y2, x3, y3)
    ans2 = ccw(x1, y1, x2, y2, x4, y4)
    ans3 = ccw(x3, y3, x4, y4, x1, y1)
    ans4 = ccw(x3, y3, x4, y4, x2, y2)

    if ans1 * ans2 == 0 and ans3 * ans4 == 0:
        return (
            min(x1, x2) <= max(x3, x4)
            and min(x3, x4) <= max(x1, x2)
            and min(y1, y2) <= max(y3, y4)
            and min(y3, y4) <= max(y1, y2)
        )
    return ans1 * ans2 <= 0 and ans3 * ans4 <= 0


n = int(sys.stdin.readline())
lines = []
parent = [i for i in range(n)]
size = [1] * n

for _ in range(n):
    lines.append(list(map(int, sys.stdin.readline().split())))

for i in range(n):
    for j in range(i + 1, n):
        if check(lines[i], lines[j]):
            union(i, j)

group_count = 0
max_size = 0

for i in range(n):
    if parent[i] == i:
        group_count += 1
        max_size = max(max_size, size[i])

print(group_count)
print(max_size)
